### DISCUSSION TITLE
Haproxy Installation not idempotent

### DISCUSSION THREAD

**Discussion by stoeps13** [HCL-TECH-SOFTWARE/connections-automation#78] [Ideas]
https://github.com/HCL-TECH-SOFTWARE/connections-automation/blob/1c22b9ed0289f61fc6955fafb29019ac45f663ec/roles/third_party/haproxy-install/tasks/configure_haproxy.yml and 
https://github.com/HCL-TECH-SOFTWARE/connections-automation/blob/d097b3274a8825657618147facf2372fe763b41d/roles/third_party/haproxy-install/tasks/download_and_build_haproxy.yml

should follow the idempotency approach and only run if anything needs a change. So downloading and building sources when the installed haproxy is not the version which will be installed.

Additionally the certificate creation should only run if not already, or when the certificate valid time is smaller a configurable time delta.

One idea would be 
```
- name:                     Check if certificates are already generated
  stat:
    path:                   /etc/haproxy/ssl/haproxy.pem
  register:                 certificate_already_exists
  ignore_errors:            true

- name:                     Check if certificate is valid
  openssl_certificate_info:
    path:                   /etc/haproxy/ssl/haproxy.pem
    valid_at:
      point_1:              "+1w"
      point_2:              "+10w"
  register:                 cert_info
  when:                     certificate_already_exists is not defined or not certificate_already_exists.stat.exists
```

The actual process generates a lot of noise of changed tasks and needs way more time on rerun as with the checks. Can't do a PR in the moment, because I started multiple workarounds, but lets discuss this. 

There are more parts in the repository where such adjustments are needed.

Regards
Christoph

**Comment by dejo1307** [comment]
I actually have this on the branch (for certificates specifically). 

Btw. I agree about idempotency. I have it on another branch currently to try to make it in a way that it still works for customers and for us. Note that we are running this over and over again (not just for Haproxy, for everything) and using it also to test every single step, where those scenarios make sense for us. Our initial guess was that customers won't care for this since they won't run it as many times against the same environment as we do, but if that is not the case, I can push on this and try to make get soon something intelligent which works for everyone.

**Reply by dejo1307** [reply]
Btw. I started fixing IBM stuff first (DB2, WAS and WAS components, TDI). Will add this on the list for sure.
